tests/metagen/typegraphs/sample/ts_upload/client.ts

```diff
@@ -0,0 +1,878 @@
+ * get {@link PlaceholderValue}s that can be used in place of node
+ * arguments. At request time, the {@link PreparedRequest} then
+ * takes an object that adheres to `T` that can then be used
+ * to replace the placeholders.
+ */
+export class PreparedArgs<T extends Record<string, unknown>> {
+  get(key: OnlyStringKeys<T>): PlaceholderValue<T[typeof key]> {
+    return new PlaceholderValue(key);
+  }
+}
+
+/**
+ * Placeholder values for use by {@link PreparedRequest}
+ */
+export class PlaceholderValue<_T> {
+  #key: string;
+  constructor(key: string) {
+    this.#key = key;
+  }
+
+  key() {
+    return this.#key;
+  }
+}
+
+export type PlaceholderArgs<T extends Record<string, unknown>> = {
+  [K in keyof T]: PlaceholderValue<T[K]>;
+};
+
+/* GraphQL section */
+
+/**
+ * Options to be used for requests performed by {@link GraphQLTransport}.
+ */
+export type GraphQlTransportOptions = Omit<RequestInit, "body"> & {
+  /**
+   * {@link fetch} implementaiton to use. Defaults to the one found in the environment
+   */
+  fetch?: typeof fetch;
+};
+
+class GqlBuilder {
+  #variables = new Map<string, NodeArgValue>();
+  #files: Map<string, File> = new Map();
+
+  #convertQueryNodeGql(
+    node: SelectNode,
+    files: ObjectPath[],
+  ) {
+    let out = node.nodeName == node.instanceName
+      ? node.nodeName
+      : `${node.instanceName}: ${node.nodeName}`;
+
+    const filesByInputKey = new Map<string, ObjectPath[]>();
+    for (const path of files) {
+      const inKey = path[0].slice(1);
+      const files = filesByInputKey.get(inKey);
+      if (files) {
+        files.push(path);
+      } else {
+        filesByInputKey.set(inKey, [path]);
+      }
+    }
+
+    const args = node.args;
+    if (args) {
+      out = `${out} (${
+        Object.entries(args)
+          .map(([key, val]) => {
+            const name = `in${this.#variables.size}`;
+            const files = filesByInputKey.get(key);
+            const object = { [key]: val.value };
+            if (files && files.length > 0) {
+              const extractedFiles = FileExtractor.extractFrom(object, files);
+              for (const [path, file] of extractedFiles.entries()) {
+                const pathInVariables = path.replace(/^\.[^\.\[]+/, `.${name}`);
+                this.#files.set(pathInVariables, file);
+              }
+            }
+            val.value = object[key]; // this might be null
+            this.#variables.set(name, val);
+            return `${key}: $${name}`;
+          })
+          .join(", ")
+      })`;
+    }
+
+    const subNodes = node.subNodes;
+    if (subNodes) {
+      // FIXME can't we have file inputs in subnodes?
+      out = `${out} { ${
+        subNodes.map((node) => this.#convertQueryNodeGql(node, [])).join(" ")
+      } }`;
+    }
+    return out;
+  }
+
+  static build(
+    typeToGqlTypeMap: Record<string, string>,
+    query: Record<string, [SelectNode, ObjectPath[]]>,
```

Comment:
You can then, while building the graphql query collect any files found in the `SelectNodes` same way we do args.